jupyter-extension/jupyterlab_pachyderm/handlers.py

```diff
@@ -443,6 +442,42 @@ async def put(self, path):
             raise e
 
 
+class ExploreDownloadHandler(BaseHandler):
+    @tornado.web.authenticated
+    async def put(self, path):
+        print(f"PUT for /download/explore{path} called")
+        try:
+            self.pfs_manager.download_file(path=path)
+        except FileExistsError:
+            raise tornado.web.HTTPError(
+                status_code=400,
+                reason=f"Downloading {Path(path).name} which already exists in the current working directory",
+            )
+        except ValueError as e:
+            raise tornado.web.HTTPError(status_code=400, reason=e)
+        except Exception as e:
+            raise tornado.web.HTTPError(status_code=500, reason=e)
+        await self.finish()
+
+
+class TestDownloadHandler(BaseHandler):
```

Comment:
What's the reason we need both download handlers?

Reply:
Since I have implemented this by adding a drop down in our file browser, we essentially get this functionality in both explore and test views for free. I think there could be use cases for wanting to download single files from a mounted datum and not the entire datum, so getting this as a bonus in addition to having it in the explore view seemed nice. To make it work, however, an additional handler is needed so that we know which manager to send the download to.

Reply:
For my own curiosity, if we wanted to, we could just route all download requests to the `pfs_manager` right? The reason why we use the `datum_manager` here as well is to maybe maintain the Test vs Explore functionality handling separation that exists. 

I think aliasing a repo in the input spec in the Test view is currently unsupported. Just one consideration for when that is implemented is accounting for aliased paths when trying to download a file.

Reply:
there is a slight difference between the two in that obtaining the repo/branch info is handled differently; each manager maintains its own records of which repos/branches are currently "mounted" which is the main reason they are handled separately. you can see this in the impl where the download logic is shared but getting the correct pfs `File` is done within the manager. 

re: the aliasing, that is probably another reason why we would keep these separate - we will want different ways of maintaining records for which repos/branches are mounted and so the mapping of a path to a pfs `File` will be different while the actual download logic should remain unchanged

Reply:
Understood, thanks!